kernel/kernel-api/src/main/java/io/delta/kernel/internal/checksum/ChecksumUtils.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (2025) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.checksum;
+
+import static io.delta.kernel.internal.actions.SingleAction.CHECKPOINT_SCHEMA;
+import static io.delta.kernel.internal.util.Preconditions.checkState;
+
+import io.delta.kernel.data.ColumnVector;
+import io.delta.kernel.data.ColumnarBatch;
+import io.delta.kernel.data.FilteredColumnarBatch;
+import io.delta.kernel.engine.Engine;
+import io.delta.kernel.internal.SnapshotImpl;
+import io.delta.kernel.internal.actions.AddFile;
+import io.delta.kernel.internal.actions.DomainMetadata;
+import io.delta.kernel.internal.replay.CreateCheckpointIterator;
+import io.delta.kernel.internal.stats.FileSizeHistogram;
+import java.io.IOException;
+import java.nio.file.FileAlreadyExistsException;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.stream.Collectors;
+
+public class ChecksumUtils {
+
+  private ChecksumUtils() {}
+
+  private static final int ADD_INDEX = CHECKPOINT_SCHEMA.indexOf("add");
+  private static final int REMOVE_INDEX = CHECKPOINT_SCHEMA.indexOf("remove");
+  private static final int DOMAIN_METADATA_INDEX = CHECKPOINT_SCHEMA.indexOf("domainMetadata");
+  private static final int ADD_SIZE_INDEX = AddFile.FULL_SCHEMA.indexOf("size");
+
+  /**
+   * Computes the state of a Delta table and writes a checksum file for the provided snapshot's
```

Comment:
I think we are loading the P&M twice since we load it on snapshot init -- I wonder if there is a way we can avoid the first instance since we do it anyways as part of this construction?

Reply:
Technically we are having same issue with checkpoint,  created https://github.com/delta-io/delta/issues/4467. I guess, all we need is not creating snapshotimpl, but just build everything with log segment.